crates/cranelift/src/compiler/component.rs

```diff
@@ -0,0 +1,170 @@
+        &self,
+        component: &Component,
+        lowering: &LowerImport,
+        types: &ComponentTypes,
+    ) -> Result<Box<dyn Any + Send>> {
+        let ty = &types[lowering.canonical_abi];
+        let isa = &*self.isa;
+        let pointer_type = isa.pointer_type();
+        let offsets = VMComponentOffsets::new(isa.pointer_bytes(), component);
+
+        let CompilerContext {
+            mut func_translator,
+            codegen_context: mut context,
+        } = self.take_context();
+
+        context.func = ir::Function::with_name_signature(
+            ir::ExternalName::user(0, 0),
+            crate::indirect_signature(isa, ty),
+        );
+
+        let mut builder = FunctionBuilder::new(&mut context.func, func_translator.context());
+        let block0 = builder.create_block();
+
+        // Start off by spilling all the wasm arguments into a stack slot to be
+        // passed to the host function.
+        let (values_vec_ptr_val, values_vec_len) =
+            self.wasm_to_host_spill_args(ty, &mut builder, block0);
+        let vmctx = builder.func.dfg.block_params(block0)[0];
+
+        // Below this will incrementally build both the signature of the host
+        // function we're calling as well as the list of arguments since the
+        // list is somewhat long.
+        let mut callee_args = Vec::new();
+        let mut host_sig = ir::Signature::new(crate::wasmtime_call_conv(isa));
+
+        let CanonicalOptions {
+            memory,
+            realloc,
+            string_encoding,
+        } = lowering.options;
+
+        // vmctx: *mut VMComponentContext
+        host_sig.params.push(ir::AbiParam::new(pointer_type));
+        callee_args.push(vmctx);
+
+        // data: *mut u8,
+        host_sig.params.push(ir::AbiParam::new(pointer_type));
+        callee_args.push(builder.ins().load(
+            pointer_type,
+            MemFlags::trusted(),
+            vmctx,
+            i32::try_from(offsets.lowering_data(lowering.index)).unwrap(),
+        ));
+
+        // memory: *mut VMMemoryDefinition
+        host_sig.params.push(ir::AbiParam::new(pointer_type));
+        callee_args.push(match memory {
+            Some(idx) => builder.ins().load(
+                pointer_type,
+                MemFlags::trusted(),
+                vmctx,
+                i32::try_from(offsets.runtime_memory(idx)).unwrap(),
+            ),
+            None => builder.ins().iconst(pointer_type, 0),
+        });
+
+        // realloc: *mut VMCallerCheckedAnyfunc
+        host_sig.params.push(ir::AbiParam::new(pointer_type));
+        callee_args.push(match realloc {
+            Some(idx) => builder.ins().load(
+                pointer_type,
+                MemFlags::trusted(),
+                vmctx,
+                i32::try_from(offsets.runtime_realloc(idx)).unwrap(),
+            ),
+            None => builder.ins().iconst(pointer_type, 0),
+        });
+
+        // string_encoding: StringEncoding
```

Comment:
Hm perhaps yeah, although this is defined in `wasmtime-environ` which doesn't have `VM*` things in general. I did leave documentation there explaining the `repr(u8)` as well, so I think I'll leave this as-is to avoid too many definitions of the same thing, but we can always tweak in the future of course.